torch/csrc/autograd/python_hook.cpp

```diff
@@ -1,42 +1,189 @@
+    check_result(grad_inputs, res, hook);
+    grad_inputs = std::move(res);
+  }
+
+  return unwrap_variables(grad_inputs.get());
 }
 
 }} // namespace torch::autograd
+
+
+static THPObjectPtr wrap_variables(const variable_list& grads)
+{
+  THPObjectPtr tuple = PyTuple_New(grads.size());
+  if (!tuple) throw python_error();
+  for (size_t i = 0; i < grads.size(); i++) {
+    THPObjectPtr grad = THPVariable_Wrap(grads[i]);
+    if (!grad) throw python_error();
+    PyTuple_SET_ITEM(tuple.get(), i, grad.release());
+  }
+  return tuple;
+}
+
+static variable_list unwrap_variables(PyObject* grads)  {
+  variable_list results(PyTuple_GET_SIZE(grads));
+  for (size_t i = 0; i < results.size(); i++) {
+    auto var = (THPVariable*)PyTuple_GET_ITEM(grads, i);
+    results[i] = var->cdata;
+  }
+  return results;
+}
+
+static void check_result(PyObject* prev, PyObject* result, PyObject* hook) {
+  if (!PyTuple_Check(result)) {
+    PyErr_Format(PyExc_TypeError, "expected tuple, but hook returned '%s'",
+        THPUtils_typename(result));
+    throw python_error();
+  }
+
+  auto prev_size = PyTuple_GET_SIZE(prev);
+  auto result_size = PyTuple_GET_SIZE(result);
+  if (prev_size != result_size) {
+    std::stringstream ss;
+    auto name = hook_name(hook);
+    ss << "backward hook '" << name << "' has returned an incorrect number ";
+    ss << "of gradients (got " << result_size << ", but expected ";
+    ss << prev_size << ")";
+    throw std::runtime_error(ss.str());
+  }
+
+  for (auto i = 0; i < prev_size; i++) {
+    check_single_result(PyTuple_GET_ITEM(prev, i), PyTuple_GET_ITEM(result, i), hook);
+  }
+}
+
+static void check_single_result(PyObject* _original, PyObject* _result, PyObject* hook) {
+  if (!PyObject_IsInstance(_result, THPVariableClass)) {
+    PyErr_Format(PyExc_TypeError, "expected Variable, but hook returned '%s'",
+        THPUtils_typename(_result));
+    throw python_error();
+  }
+
+  auto& original = *((THPVariable*)_original)->cdata->data;
+  auto& result = *((THPVariable*)_result)->cdata->data;
+
+  if (original.type() != result.type()) {
+    std::stringstream ss;
+    auto name = hook_name(hook);
+    ss << "backward hook '" << name << "' has changed the type of grad_input (";
+    ss << "was " << thpp::toString(original.type()) << " got ";
+    ss << thpp::toString(result.type()) << ")";
+    throw std::runtime_error(ss.str());
+  }
+
+  if (original.isCuda() != result.isCuda()) {
+    std::stringstream ss;
+    auto name = hook_name(hook);
+    ss << "backward hook '" << name << "' has changed the type of grad_input";
+    if (original.isCuda()) {
+      ss << " (was CUDA tensor got CPU tensor)";
+    } else {
+      ss << " (was CPU tensor got CUDA tensor)";
+    }
+    throw std::runtime_error(ss.str());
+  }
+
+  if (original.sizes() != result.sizes()) {
+    std::stringstream ss;
+    auto name = hook_name(hook);
+    ss << "backward hook '" << name << "' has changed the size of grad_input";
```

Comment:
Probably... these are the same error messages that were in python_function.cpp previously.